### PR TITLE
[JVM IR] Fix constant folding to use basic types always.

### PR DESCRIPTION
**Pull request by madsager**
Fixes KT-46540.

### CODE REVIEW

**Comment by madsager** [comment]
@pyos has this change out that would also fix this: https://github.com/JetBrains/kotlin/pull/3747 by making codegen deal with `IrConst`s that have a type that do not correspond to an asm primitive type. For the constant folding, this change is simple and works. @pyos's change would allow more flexible use of `IrConst`s and remove the implicit assumption that their type must map to the right asm primitive type.

File: compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/optimizations/FoldConstantLowering.kt [hunk 1]
```diff
@@ -107,16 +107,16 @@ class FoldConstantLowering(
     private fun buildIrConstant(startOffset: Int, endOffset: Int, type: IrType, v: Any?): IrConst<*> {
         val constType = type.makeNotNull()
```
Comment (udalov) [inline compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/optimizations/FoldConstantLowering.kt:108]: This is now used only once below in `constType.isString()`, and probably can be replaced with `type.isStringClassType()`

**Comment by udalov** [comment]
@romanart @homuroll Please verify that it's OK for JS/Native

File: compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/optimizations/FoldConstantLowering.kt (hunk 1 above)
Comment (madsager) [reply compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/optimizations/FoldConstantLowering.kt:108]: Thanks. Done.

**Review by romanart — APPROVED** [review]
LGTM

**Comment by homuroll** [comment]
LGTM

**Review by homuroll — APPROVED** [review]

**Comment by udalov** [comment]
Thank you!

**Pull request merged**